agent/lib_aws_sdk_php.c

```diff
@@ -0,0 +1,165 @@
+}
+
+/*
+ * Any wrapped function pointing here will automatically have it's class
+ * name created as a supportability metric.
+ */
+NR_PHP_WRAPPER(nr_aws_create_metric) {
+  if (NULL != wraprec) {
+    lib_aws_sdk_php_add_supportability_metric(wraprec->classname);
+  }
+
+  NR_PHP_WRAPPER_CALL;
+}
+NR_PHP_WRAPPER_END
+
+NR_PHP_WRAPPER(nr_aws_version) {
+  (void)wraprec;
+
+  lib_aws_sdk_php_handle_version();
+  NR_PHP_WRAPPER_CALL;
+}
+NR_PHP_WRAPPER_END
+
+/*
+ * The ideal file to begin immediate detection of the aws-sdk is:
+ * aws-sdk-php/src/functions.php
+ * Unfortunately, Php8.2+ and composer autoload leads to the
+ * file being optimized directly and not loaded.
+ *
+ * Options considered:
+ *
+ * 1. for PHP8.2, and only optimizable libraries, when encountering autoload.php
+ * files, ask the file what includes it added and check against only the
+ * optimizable library. Small overhead incurred when encountering an autoload
+ * file, but detects aws-sdk-php immediately before any sdk code executes
+ * (changes needed for this are detailed in the original PR)
+ * 2. use a file that gets called later and only when AwsClient.php file file is
+ * called. It's called later and we'll miss some instrumentation, but if we're
+ * only ever going to be interested in Client calls anyway, maybe that's ok?
+ * Doesn't detect Sdk.php (optimized out) so when customers only use that or
+ * when they use it first, we will not instrument it. This only detects when a
+ * Client is called to use a service so potentially misses out on other
+ * instrumentation and misses out when customers use the aws-sdk-php but use
+ * non-SDK way to interact with the service (possibly with redis/memcached).
+ * This way is definitely the least complex and lowest overhead and less
+ * complexity means lower risk as well.
+ * 3. Directly add the wrappers to the hash map. With potentially 50ish clients
+ * to wrap, this will add overhead to every hash map lookup. Currently
+ * implemented option is 2, use the AwsClient.php as this is our main focus.
+ * This means until a call to an Aws/AwsClient function,
+ * all calls including aws\sdk calls are ignored.  Version detection will be
+ * tied to Aws/ClientResolver::_apply_user_agent which is ALWAYS called when
+ * dealing with aws clients.  It will not be computed from
+ * Aws/Sdk::__constructor which would at best be duplicate info and worst would
+ * never be ignored until a client is called.
+ */
+void nr_aws_sdk_php_enable() {
+  /* This is will guarantee we can extract the version. */
+  nr_php_wrap_user_function(NR_PSTR("Aws\\ClientResolver::_apply_user_agent"),
+                            nr_aws_version);
+  /* This is the base of all other Clients */
+  nr_php_wrap_user_function(NR_PSTR("Aws\\AwsClient::__construct"),
+                            nr_aws_create_metric);
+  /* The Sdk class */
+  nr_php_wrap_user_function(NR_PSTR("Aws\\Sdk::__construct"),
+                            nr_aws_create_metric);
```

Comment:
Why these two methods need to be wrapped? I thought we're interested in generating metrics for:
- aws-sdk-php@v3 package to understand usage of that package
- AWS services used by PHP applications to understand usage of AWS services by PHP applications

First objective is achieved (mostly - see [this comment](https://github.com/newrelic/newrelic-php-agent/pull/932/files#r1681801550)) with wrapping `Aws\\ClientResolver::_apply_user_agent` - `nr_aws_version` generates `Supportability/PHP/package/aws/aws-sdk-php/3/detected` metric.

Implementation of the second objective for S3 service is demonstrated with wrapping `Aws\\S3\\S3Client::__construct` - `nr_aws_create_metric` tries to generate this metric (however it would benefit from [my other suggestion](https://github.com/newrelic/newrelic-php-agent/pull/932/files#r1681470624)). What value does wrapping `Aws\AwsClient::__construct` and `Aws\Sdk::__construct` give? The metric generated by `nr_aws_create_metric` does not provide the name of AWS service - just the name of the class (`Aws\AwsClient` and `Aws\Sdk`).

Reply:
Please see: [7d98b70](https://github.com/newrelic/newrelic-php-agent/pull/932/commits/7d98b70ca2e9ada5dc79739fefa63e535ae6220d)

Both are removed as unneeded.